### ISSUE TITLE
Builcache fails to relocate xz

### ISSUE THREAD

**Issue by JavierCVilla** [spack/spack#12226]
Spack buildcache fails to relocate the `xz` package.

### Steps to reproduce the issue

1. Set up the inital Spack instance:

```console
$ git clone https://github.com/spack/spack
$ source spack/share/spack/setup-env.sh
```

2. Install `xz` package:

```console
$ spack install xz
```
3. Create binary for xz:

```console
$ spack buildcache create -au -d /tmp/localbuildcache xz
```

4. Set up a new spack instance in a different location:

```console
$ mkdir secondspack
$ cd secondspack
$ git clone https://github.com/spack/spack
$ source spack/share/spack/setup-env.sh
```

5. Add local buildcache:

```console
$ spack mirror add local /tmp/localbuildcache/
$ spack mirror list
local    file:///tmp/localbuildcache/
```

6. Install `xz` from buildcache:

```console
$ spack buildcache install -u xz
==> Finding buildcaches in /tmp/localbuildcache//build_cache
==> Fetching file:///tmp/localbuildcache//build_cache/linux-centos7-x86_64-gcc-4.8.5-xz-5.2.4-agv5zlv7phi4ye3evlpv3t4ypgve4bpe.spec.yaml
######################################################################## 100.0%
==> buildcache spec(s) matching xz 

==> Fetching file:///tmp/localbuildcache//build_cache/linux-centos7-x86_64/gcc-4.8.5/xz-5.2.4/linux-centos7-x86_64-gcc-4.8.5-xz-5.2.4-agv5zlv7phi4ye3evlpv3t4ypgve4bpe.spack
######################################################################## 100.0%
==> Installing buildcache for spec xz@5.2.4%gcc@4.8.5 arch=linux-centos7-x86_64
==> Relocating package from
  /path/spack/opt/spack to /path/secondspack/spack/opt/spack.
==> Installing patchelf
==> Fetching file:///tmp/localbuildcache/patchelf/patchelf-0.9.tar.gz
curl: (37) Couldn't open file /tmp/localbuildcache/patchelf/patchelf-0.9.tar.gz
==> Fetching from file:///tmp/localbuildcache/patchelf/patchelf-0.9.tar.gz failed.
==> Fetching http://nixos.org/releases/patchelf/patchelf-0.9/patchelf-0.9.tar.gz
######################################################################## 100.0%
==> Staging archive: /path/secondspack/spack/var/spack/stage/patchelf-0.9-ufwuiecywz2zepil46ifqlophnhwatbi/patchelf-0.9.tar.gz
==> Created stage in /path/secondspack/spack/var/spack/stage/patchelf-0.9-ufwuiecywz2zepil46ifqlophnhwatbi
==> No patches needed for patchelf
==> Building patchelf [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed patchelf
  Fetch: 0.41s.  Build: 8.59s.  Total: 9.01s.
[+] /path/secondspack/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/patchelf-0.9-ufwuiecywz2zepil46ifqlophnhwatbi
==> Error: 
```

### Error Message

The error message reported is empty (`==> Error:`) :

```console
$ spack -d --stacktrace buildcache install -u xz

lib/spack/spack/util/executable.py:159 ==> [2019-08-01-16:45:07.830899] '/path/secondspack/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/patchelf-0.9-ufwuiecywz2zepil46ifqlophnhwatbi/bin/patchelf' '--force-rpath' '--set-rpath' '/path/secondspack/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/xz-5.2.4-agv5zlv7phi4ye3evlpv3t4ypgve4bpe/lib:/path/secondspack/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/xz-5.2.4-agv5zlv7phi4ye3evlpv3t4ypgve4bpe/lib64' '/tmp/sftnight/tmp6llrJZ/xz-5.2.4-agv5zlv7phi4ye3evlpv3t4ypgve4bpe/bin/xz'
lib/spack/spack/binary_distribution.py:578 ==> [2019-08-01-16:45:08.095875] Error: 
Traceback (most recent call last):
  File "/path/secondspack/spack/lib/spack/spack/main.py", line 697, in main
    return _invoke_command(command, parser, args, unknown)
  File "/path/secondspack/spack/lib/spack/spack/main.py", line 447, in _invoke_command
    return_val = command(parser, args)
  File "/path/secondspack/spack/lib/spack/spack/cmd/buildcache.py", line 557, in buildcache
    args.func(args)
  File "/path/secondspack/spack/lib/spack/spack/cmd/buildcache.py", line 351, in installtarball
    install_tarball(match, args)
  File "/path/secondspack/spack/lib/spack/spack/cmd/buildcache.py", line 370, in install_tarball
    args.unsigned, args.force)
  File "/path/secondspack/spack/lib/spack/spack/binary_distribution.py", line 578, in extract_tarball
    tty.die(e)
  File "/path/secondspack/spack/lib/spack/llnl/util/tty/__init__.py", line 225, in die
    sys.exit(1)
SystemExit: 1

```
 
The issue seems to come from the [`assert`](https://github.com/spack/spack/blob/cc4094bb9b6b228ef9c3821e9e76361df50b7f94/lib/spack/spack/relocate.py#L312) in the `replace_prefix_bin` that checks the length of the binary before and after applying the regex substitution.

The binary that triggers the issue is: `xz-5.2.4/bin/lzmainfo`. Before applying the substitution the lenght is:

```
(Pdb) print(len(data))
32376
```

And after the substitution it almost doubles:

```
(Pdb) print(len(data))
64603
```

### Information on your system

 1. Platform: Centos7, default compiler (gcc-4.8.5)

```
LSB Version:	:core-4.1-amd64:core-4.1-noarch
Distributor ID:	CentOS
Description:	CentOS Linux release 7.6.1810 (Core) 
Release:	7.6.1810
Codename:	Core
```

 2. No additional configuration.

**Comment by JavierCVilla** [comment]
@gartung

**Comment by gartung** [comment]
@JavierCVilla I opened this PR to use an exception instead of an assert:
https://github.com/spack/spack/pull/12227

**Comment by JavierCVilla** [comment]
@gartung thanks! That will definitely make it easier to understand. 

Regarding the error itself, any idea why the size can be different?

**Comment by gartung** [comment]
I copied the padding algorithm from conda and it worked for relocating git and tcl. I am trying to debug what happened with xz.

**Comment by gartung** [comment]
Was the buildcache created on another machine? Was the original install path longer or shorter than original install path? I am unable to reproduce the problem on my linux desktop going from a longer path to a shorter path. Conda adds padding to the name of the build directory.

**Comment by JavierCVilla** [comment]
I just tried on the same machine and I was going from a shorter to a longer path.

**Comment by gartung** [comment]
That would cause problems. I am going to skip the replacement in that case.

**Comment by gartung** [comment]
@JavierCVilla I just merged #12227 to the head of develop. I tested that it fixes the problem of installing a buildcache created with an original install path shorter than the new install path. It will now issue a warning instead of trying to do the binary text replacement with padding. On the off chance that someone calls the function directly an exception will be thrown with additional info.
